src/Client.ts

```diff
@@ -195,23 +202,34 @@ export default class Client {
     return client.legacyKeys.encode()
   }
 
-  async init(options: ClientOptions): Promise<void> {
+  private async init(options: ClientOptions): Promise<void> {
     options.codecs.forEach((codec) => {
       this.registerCodec(codec)
     })
     this._maxContentSize = options.maxContentSize
-    await this.publishUserContact()
+    await this.ensureUserContactPublished(options.publishLegacyContact)
   }
 
   // gracefully shut down the client
   async close(): Promise<void> {
     return undefined
   }
 
-  // publish the key bundle into the contact topic
-  // WARNING: temporarily public to allow testing negotiated topics
-  // TODO: make private again asap
-  async publishUserContact(legacy = true): Promise<void> {
+  private async ensureUserContactPublished(legacy = false): Promise<void> {
+    const bundle = await getUserContactFromNetwork(this.apiClient, this.address)
+    if (
+      bundle &&
+      bundle instanceof SignedPublicKeyBundle &&
+      this.keys.getPublicKeyBundle().equals(bundle)
+    ) {
+      return
+    }
+    this.publishUserContact(legacy)
+  }
+
+  // PRIVATE: publish the key bundle into the contact topic
+  // left public for testing purposes
+  async publishUserContact(legacy = false): Promise<void> {
```

Comment:
The mixed v1/v2 test cases still need to be able to republish the contact, so I have to leave this open.